### DISCUSSION TITLE
Preamble is 1010..., not 0101... as specified in the USB PD spec

### DISCUSSION THREAD

**Discussion by eliasnaur** [manuelbl/usb-pd-arduino#19] [General]
Thank you for this repository of code for communicating with USB power supplies. I working on my own implementation, but so far failed to make the power supply recognize my GoodCRC and Request messages.

I'm looking at your code as reference, and ran into a bug or misunderstanding regarding the PD packet preamble.

In the encoding of the preamble,

https://github.com/manuelbl/usb-pd-arduino/blob/a9765e7559befb7940fb51d3a893524b16f0dbb8/src/PDMessageEncoder.cpp#L90-L92

there are 16 chunks of "1010" (in BMC). However, the USB PD specification section 5.6.1.1 states that "The Preamble Shall consist of a 64-bit sequence of alternating 0s and 1s. The Preamble Shall start with a "0" and
Shall end with a "1"." This pattern is also what I see when receiving Source_Capabilities messages in my own implementation.

As such, I had expected

```C
    // preamble
    for (int i = 0; i < 16; i++)
        buffer[i] = 0b00101101; // "0101" in BMC.
```

Your code obviously works, and mine doesn't. What did I miss?

**Comment by manuelbl** [comment]
There are two layers to the protocol as shown in this illustration from the USB PD specification (revision 3.1, version 1.8)

<img width="408" alt="BMC example" src="https://github.com/user-attachments/assets/db60e90b-2ac1-4d6f-a64b-4fd9d9305df4">

At the data layer, the preamble is indeed 010101... However, at the BMC level (signal you would observe on the oscilloscope), 0 becomes a full clock cycle at the same level and 1 becomes a transition from 0 to 1 (or vice versa) in the middle of the cycle, plus a transition at the end of each cycle for both cases.

In my code, 1 bit is used for half a clock cycle. So the pattern 0b10110100 represent 4 data bits at the BMC level (and in little-endian representation; so it's transmitted from right to left). At the data layer it would be 0 1 0 1. This pattern is repeated 16 times to make up the 64-bit sequence. The resulting BMC bits are transmitted at twice the nominal clock rate.

I don't know your code and if you need the pattern at the data layer or the BMC layer. But it would explain the difference.

**Comment by eliasnaur** [comment]
Thank you very much for the detailed answer. I had missed you read the data in little-endian such that your 0b10110100 pattern indeed matches the preamble.

I also got my own implementation to work. The issue was that "USB PD protocol is very timing sensitive" :)
